sdk/nodejs/cmd/pulumi-language-nodejs/main.go

```diff
@@ -807,16 +810,106 @@ func (host *nodeLanguageHost) InstallDependencies(
 
 	stdout.Write([]byte("Installing dependencies...\n\n"))
 
-	_, err = npm.Install(ctx, req.Info.ProgramDirectory, false /*production*/, stdout, stderr)
+	workspaceRoot := req.Info.ProgramDirectory
+	newWorkspaceRoot, err := findWorkspaceRoot(req.Info.ProgramDirectory)
+	if err != nil {
+		// If we are not in a npm/yarn workspace, we will keep the current directory as the root.
+		if !errors.Is(err, errNotInWorkspace) {
+			return fmt.Errorf("failure while trying to find workspace root: %w", err)
+		}
+	} else {
+		stdout.Write([]byte(fmt.Sprintf("Detected workspace root at %s\n", newWorkspaceRoot)))
+		workspaceRoot = newWorkspaceRoot
+	}
+
+	_, err = npm.Install(ctx, workspaceRoot, false /*production*/, stdout, stderr)
 	if err != nil {
-		return fmt.Errorf("npm install failed: %w", err)
+		return fmt.Errorf("dependency installation failed: %w", err)
 	}
 
 	stdout.Write([]byte("Finished installing dependencies\n\n"))
 
 	return closer.Close()
 }
 
+func findWorkspaceRoot(programDirectory string) (string, error) {
```

Comment:
@Frassle looks like I might need this function inside ProgramTest in another PR, where would I put this function to share it?

Reply:
Move it up to a .go file in sdk/nodejs. That is (somewhat surprisingly) part of the go SDK module so ProgramTest can link to it.